// bandcamp-fetch/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cache = exports.CacheDataType = exports.AutocompleteAPI = exports.SearchAPI = exports.TrackAPI = exports.TagAPI = exports.ShowAPI = exports.ImageAPI = exports.FanAPI = exports.DiscoveryAPI = exports.BandAPI = exports.ArticleAPI = exports.AlbumAPI = void 0;
const AlbumAPI_1 = __importStar(require("./lib/album/AlbumAPI"));
const ArticleAPI_1 = __importStar(require("./lib/article/ArticleAPI"));
const BandAPI_1 = __importStar(require("./lib/band/BandAPI"));
const DiscoveryAPI_1 = __importStar(require("./lib/discovery/DiscoveryAPI"));
const FanAPI_1 = __importStar(require("./lib/fan/FanAPI"));
const ImageAPI_1 = __importStar(require("./lib/image/ImageAPI"));
const ShowAPI_1 = __importStar(require("./lib/show/ShowAPI"));
const TagAPI_1 = __importStar(require("./lib/tag/TagAPI"));
const TrackAPI_1 = __importStar(require("./lib/track/TrackAPI"));
const SearchAPI_1 = __importStar(require("./lib/search/SearchAPI"));
const AutocompleteAPI_1 = __importStar(require("./lib/autocomplete/AutocompleteAPI"));
const Limiter_1 = __importDefault(require("./lib/utils/Limiter"));
const Cache_1 = require("./lib/utils/Cache");
var AlbumAPI_2 = require("./lib/album/AlbumAPI");
Object.defineProperty(exports, "AlbumAPI", { enumerable: true, get: function () { return __importDefault(AlbumAPI_2).default; } });
var ArticleAPI_2 = require("./lib/article/ArticleAPI");
Object.defineProperty(exports, "ArticleAPI", { enumerable: true, get: function () { return __importDefault(ArticleAPI_2).default; } });
var BandAPI_2 = require("./lib/band/BandAPI");
Object.defineProperty(exports, "BandAPI", { enumerable: true, get: function () { return __importDefault(BandAPI_2).default; } });
var DiscoveryAPI_2 = require("./lib/discovery/DiscoveryAPI");
Object.defineProperty(exports, "DiscoveryAPI", { enumerable: true, get: function () { return __importDefault(DiscoveryAPI_2).default; } });
var FanAPI_2 = require("./lib/fan/FanAPI");
Object.defineProperty(exports, "FanAPI", { enumerable: true, get: function () { return __importDefault(FanAPI_2).default; } });
var ImageAPI_2 = require("./lib/image/ImageAPI");
Object.defineProperty(exports, "ImageAPI", { enumerable: true, get: function () { return __importDefault(ImageAPI_2).default; } });
var ShowAPI_2 = require("./lib/show/ShowAPI");
Object.defineProperty(exports, "ShowAPI", { enumerable: true, get: function () { return __importDefault(ShowAPI_2).default; } });
var TagAPI_2 = require("./lib/tag/TagAPI");
Object.defineProperty(exports, "TagAPI", { enumerable: true, get: function () { return __importDefault(TagAPI_2).default; } });
var TrackAPI_2 = require("./lib/track/TrackAPI");
Object.defineProperty(exports, "TrackAPI", { enumerable: true, get: function () { return __importDefault(TrackAPI_2).default; } });
var SearchAPI_2 = require("./lib/search/SearchAPI");
Object.defineProperty(exports, "SearchAPI", { enumerable: true, get: function () { return __importDefault(SearchAPI_2).default; } });
var AutocompleteAPI_2 = require("./lib/autocomplete/AutocompleteAPI");
Object.defineProperty(exports, "AutocompleteAPI", { enumerable: true, get: function () { return __importDefault(AutocompleteAPI_2).default; } });
__exportStar(require("./lib/album/AlbumAPI"), exports);
__exportStar(require("./lib/discovery/DiscoveryAPI"), exports);
__exportStar(require("./lib/image/ImageAPI"), exports);
__exportStar(require("./lib/article/ArticleAPI"), exports);
__exportStar(require("./lib/band/BandAPI"), exports);
__exportStar(require("./lib/fan/FanAPI"), exports);
__exportStar(require("./lib/show/ShowAPI"), exports);
__exportStar(require("./lib/tag/TagAPI"), exports);
__exportStar(require("./lib/track/TrackAPI"), exports);
__exportStar(require("./lib/album/AlbumAPI"), exports);
__exportStar(require("./lib/search/SearchAPI"), exports);
__exportStar(require("./lib/autocomplete/AutocompleteAPI"), exports);
__exportStar(require("./lib/types/Album"), exports);
__exportStar(require("./lib/types/Article"), exports);
__exportStar(require("./lib/types/Artist"), exports);
__exportStar(require("./lib/types/Discovery"), exports);
__exportStar(require("./lib/types/Fan"), exports);
__exportStar(require("./lib/types/Image"), exports);
__exportStar(require("./lib/types/Label"), exports);
__exportStar(require("./lib/types/MediaKind"), exports);
__exportStar(require("./lib/types/Show"), exports);
__exportStar(require("./lib/types/Tag"), exports);
__exportStar(require("./lib/types/Track"), exports);
__exportStar(require("./lib/types/UserKind"), exports);
__exportStar(require("./lib/types/Search"), exports);
__exportStar(require("./lib/types/Autocomplete"), exports);
var Cache_2 = require("./lib/utils/Cache");
Object.defineProperty(exports, "CacheDataType", { enumerable: true, get: function () { return Cache_2.CacheDataType; } });
class Cache {
    static clear(type) {
        Cache_1.Cache.clear(type);
    }
    static setTTL(type, ttl) {
        Cache_1.Cache.setTTL(type, ttl);
    }
    static setMaxPages(maxPages) {
        Cache_1.Cache.setMaxEntries(Cache_1.CacheDataType.Page, maxPages);
    }
}
exports.Cache = Cache;
exports.default = {
    album: AlbumAPI_1.default,
    track: TrackAPI_1.default,
    discovery: DiscoveryAPI_1.default,
    image: ImageAPI_1.default,
    band: BandAPI_1.default,
    tag: TagAPI_1.default,
    article: ArticleAPI_1.default,
    show: ShowAPI_1.default,
    fan: FanAPI_1.default,
    search: SearchAPI_1.default,
    autocomplete: AutocompleteAPI_1.default,
    limiter: {
        album: AlbumAPI_1.LimiterAlbumAPI,
        track: TrackAPI_1.LimiterTrackAPI,
        discovery: DiscoveryAPI_1.LimiterDiscoveryAPI,
        image: ImageAPI_1.LimiterImageAPI,
        band: BandAPI_1.LimiterBandAPI,
        tag: TagAPI_1.LimiterTagAPI,
        article: ArticleAPI_1.LimiterArticleAPI,
        show: ShowAPI_1.LimiterShowAPI,
        fan: FanAPI_1.LimiterFanAPI,
        search: SearchAPI_1.LimiterSearchAPI,
        autocomplete: AutocompleteAPI_1.LimiterAutocompleteAPI,
        updateSettings: Limiter_1.default.updateSettings.bind(Limiter_1.default)
    },
    cache: Cache
};
