// bandcamp-fetch/dist/mjs/lib/utils/Fetch.js
import { URL } from 'url';
import fetch from 'node-fetch';
import { Cache, CacheDataType } from './Cache';
export var FetchMethod;
(function (FetchMethod) {
    FetchMethod["GET"] = "GET";
    FetchMethod["POST"] = "POST";
})(FetchMethod || (FetchMethod = {}));
export function fetchPage(url, jsonResponse, method, payload) {
    if (jsonResponse === undefined) {
        jsonResponse = false;
    }
    return Cache.getOrSet(CacheDataType.Page, getCacheKey(url, jsonResponse, payload), async () => {
        if (method === undefined) {
            method = FetchMethod.GET;
        }
        let response;
        if (method === FetchMethod.GET) {
            const urlObj = new URL(url);
            if (payload) {
                for (const [key, value] of Object.entries(payload)) {
                    urlObj.searchParams.set(key, value);
                }
            }
            try {
                response = await fetch(urlObj.toString());
            }
            catch (error) {
                throw new FetchError(error);
            }
        }
        else {
            const init = {
                method: 'POST',
                body: payload ? JSON.stringify(payload) : undefined,
                headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
            };
            try {
                response = await fetch(url, init);
            }
            catch (error) {
                throw new FetchError(error);
            }
        }
        if (response.status === 429) {
            throw new FetchError({
                message: '429 Too Many Requests',
                code: 429
            });
        }
        if (jsonResponse) {
            return response.json();
        }
        return response.text();
    });
}
function getCacheKey(url, jsonResponse, payload) {
    return url + (jsonResponse ? ':json' : ':html') +
        (payload ? `:${JSON.stringify(payload)}` : '');
}
export class FetchError extends Error {
    constructor(payload) {
        super();
        if (payload?.message) {
            this.message = payload.message;
        }
        if (payload?.code) {
            this.code = payload.code;
        }
        if (payload?.stack) {
            this.stack = payload.stack;
        }
    }
}
