// bandcamp-fetch/dist/cjs/lib/track/TrackAPI.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LimiterTrackAPI = void 0;
const ImageAPI_1 = __importDefault(require("../image/ImageAPI"));
const Fetch_1 = require("../utils/Fetch");
const Limiter_1 = __importDefault(require("../utils/Limiter"));
const TrackInfoParser_1 = __importDefault(require("./TrackInfoParser"));
class TrackAPI {
    static async getInfo(params) {
        const imageConstants = await ImageAPI_1.default.getConstants();
        const opts = {
            trackUrl: params.trackUrl,
            imageBaseUrl: imageConstants.baseUrl,
            albumImageFormat: await ImageAPI_1.default.getFormat(params.albumImageFormat, 9),
            artistImageFormat: await ImageAPI_1.default.getFormat(params.artistImageFormat, 21),
            includeRawData: params.includeRawData !== undefined ? params.includeRawData : false
        };
        const html = await (0, Fetch_1.fetchPage)(params.trackUrl);
        return TrackInfoParser_1.default.parseInfo(html, opts);
    }
}
exports.default = TrackAPI;
class LimiterTrackAPI extends TrackAPI {
    static async getInfo(params) {
        return Limiter_1.default.schedule(() => super.getInfo(params));
    }
}
exports.LimiterTrackAPI = LimiterTrackAPI;
