// bandcamp-fetch/dist/cjs/lib/band/BandAPI.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LimiterBandAPI = void 0;
const ImageAPI_1 = __importDefault(require("../image/ImageAPI"));
const Fetch_1 = require("../utils/Fetch");
const Limiter_1 = __importDefault(require("../utils/Limiter"));
const Parse_1 = require("../utils/Parse");
const BandInfoParser_1 = __importDefault(require("./BandInfoParser"));
const DiscographyParser_1 = __importDefault(require("./DiscographyParser"));
const LabelArtistsParser_1 = __importDefault(require("./LabelArtistsParser"));
class BandAPI {
    static async getDiscography(params) {
        const imageConstants = await ImageAPI_1.default.getConstants();
        const opts = {
            imageBaseUrl: imageConstants.baseUrl,
            bandUrl: params.bandUrl,
            imageFormat: await ImageAPI_1.default.getFormat(params.imageFormat, 9)
        };
        const html = await (0, Fetch_1.fetchPage)((0, Parse_1.normalizeUrl)('music', params.bandUrl));
        return DiscographyParser_1.default.parseDiscography(html, opts);
    }
    static async getInfo(params) {
        const opts = {
            bandUrl: params.bandUrl,
            imageFormat: await ImageAPI_1.default.getFormat(params.imageFormat, 21)
        };
        const url = this.getUrl(params.bandUrl);
        const html = await (0, Fetch_1.fetchPage)(url);
        const result = BandInfoParser_1.default.parseInfo(html, opts);
        // Return if result is complete
        if (this.isInfoComplete(result)) {
            return result;
        }
        // Info lacking name or label (for artist) - try getting them from music page
        const musicUrl = this.getUrl(params.bandUrl, 'music');
        const musicHtml = await (0, Fetch_1.fetchPage)(musicUrl);
        const info = BandInfoParser_1.default.parseInfo(musicHtml, opts);
        this.fillInfo(result, info);
        // Return if result is complete
        if (this.isInfoComplete(result)) {
            return result;
        }
        // Info is still lacking name or label (for artist) - last try with fetching
        // From discog's first album or track
        const discogItems = await this.getDiscography(params);
        if (discogItems[0]) {
            const url = discogItems[0]?.url;
            if (url) {
                const html = await (0, Fetch_1.fetchPage)(url);
                const info = BandInfoParser_1.default.parseInfo(html, opts);
                this.fillInfo(result, info);
            }
        }
        if (result.url === null) {
            result.url = params.bandUrl;
        }
        return result;
    }
    static async getLabelArtists(params) {
        const opts = {
            labelUrl: params.labelUrl,
            imageFormat: await ImageAPI_1.default.getFormat(params.imageFormat)
        };
        const html = await (0, Fetch_1.fetchPage)((0, Parse_1.normalizeUrl)('artists', params.labelUrl));
        return LabelArtistsParser_1.default.parseLabelArtists(html, opts);
    }
    /**
     * @internal
     */
    static getUrl(artistOrLabelUrl, path, labelId) {
        let url = path ? (0, Parse_1.normalizeUrl)(path, artistOrLabelUrl) : artistOrLabelUrl;
        if (labelId) {
            url += `/?label=${encodeURIComponent(labelId)}`;
        }
        return url;
    }
    /**
     * @internal
     */
    static isInfoComplete(data) {
        return data.name && data.url &&
            (data.type === 'label' || data.label);
    }
    /**
     * @internal
     */
    static fillInfo(target, src) {
        if (target.name === null) {
            target.name = src.name;
        }
        if (target.url === null) {
            target.url = src.url;
        }
        if (target.type === 'artist' && src.type === 'artist' && !target.label && src.label) {
            target.label = src.label;
        }
        return target;
    }
}
exports.default = BandAPI;
class LimiterBandAPI extends BandAPI {
    static async getDiscography(params) {
        return Limiter_1.default.schedule(() => super.getDiscography(params));
    }
    static async getInfo(params) {
        return Limiter_1.default.schedule(() => super.getInfo(params));
    }
    static async getLabelArtists(params) {
        return Limiter_1.default.schedule(() => super.getLabelArtists(params));
    }
}
exports.LimiterBandAPI = LimiterBandAPI;
